### ISSUE TITLE
Running LLVM IR causes segfault on macOS aarch64 but not on x86_64

### ISSUE THREAD

**Issue by hameerabbasi** [llvm/llvm-project#90025]
TL;DR: Identical LLVM IR produced on both platforms, only one platform consistently crashes. Seems like an ABI thing as passing values in and adding print statements shows.

Hello, I have a version of the LLVM toolchain at commit `c2a98fdeb3aede1a8db492a6ea30f4fa85b60edc` compiled on both an aarch64 and x86_64 Mac with the following commands:

```bash
cmake -G Ninja ../llvm \
   -DLLVM_ENABLE_PROJECTS="mlir;llvm;clang;lld" \
   -DLLVM_BUILD_EXAMPLES=ON \
   -DLLVM_TARGETS_TO_BUILD="Native" \
   -DCMAKE_BUILD_TYPE=Release \
   -DLLVM_ENABLE_ASSERTIONS=ON
ninja
```

## What I've found so far
I'm attaching the full reproducer here: [sddmm_repro.zip](https://github.com/llvm/llvm-project/files/15108710/sddmm_repro.zip)

After running `./compile.sh`, `sddmm_opt.mlir` has the following function signature (reorganized for style):

```ll
llvm.func @_mlir_ciface_sddmm_kernel(
    %arg0: !llvm.ptr, // to output struct holding memref<?xi64>, memref<?xi64>, memref<?xf64>, type of %arg6
    %arg1: !llvm.ptr, // to memref<?x?xf64>
    %arg2: !llvm.ptr, // to memref<?x?xf64>
    %arg3: !llvm.ptr, // to memref<?xi64>
    %arg4: !llvm.ptr, // to memref<?xi64>
    %arg5: !llvm.ptr, // to memref<?xf64>
    %arg6: !llvm.struct<(array<2 x i64>, array<3 x i64>)>) attributes {llvm.emit_c_interface} {
  ...
}
```

By reading the generated function definition after adding some sparse_tensor.print ops to the original file and linking with the right `*.o` files, I found that on aarch64, the last argument, `%arg6` isn’t getting passed in properly, it seems to have some garbage values. Since this contains some loop endpoints, it, of course, segfaults.

Changing to a simpler kernel that just prints and returns a `%args` in `sddmm.mlir` reveals that the last field contained in `%arg0` with the same type as `%arg6` reveals that returning isn’t an issue – It copies the right values over, so passing the struct into MLIR is somehow borked.

So, looking at the part of the code defining the input struct corresponding to `%arg6`, we see:

```c++
struct Shapes {
  intptr_t sizes[2];
  intptr_t lengths[3];
};
```

which contains no pointers or “shared” data, so that shouldn’t be an issue either.

Digging deeper, it seems the `sddmm_opt.ll` file generated is identical across both machines, and contains nothing platform-specific other than the alignment (8 bytes for both systems). Here is the LLVM generated function signature:

```ll
define void @_mlir_ciface_sddmm_kernel(ptr %0, ptr %1, ptr %2, ptr %3, ptr %4, ptr %5, { [2 x i64], [3 x i64] } %6) {
  ...
}
```

## To run the repoducer
To reproduce, out of the full set of files (included for posterity), you only need three files besides your LLVM build:
1. `compile.sh` 
2. `sddmm.mlir`
3. Either `sddmm_test.cpp` or `sddmm_test.py`

Just edit the paths (line 6 of `compile.sh` and line 121 of `sddmm_test.py`) and run either `./a.out` or `python sddmm_test.py`.

Edit 2024.04.25, 11:15 UTC: I had a slight bug in my C++ code where the argument order was wrong, I've uploaded the right reproducer now.

**Comment by hameerabbasi** [comment]
Okay, so I just debugged this with one of my colleagues who's more knowledgeable in this kind of stuff. The issue, from our investigations, lies in the C interface emission.

> After running `./compile.sh`, `sddmm_opt.mlir` has the following function signature (reorganized for style):
> 
> ```llvm
> llvm.func @_mlir_ciface_sddmm_kernel(
>     %arg0: !llvm.ptr, // to output struct holding memref<?xi64>, memref<?xi64>, memref<?xf64>, type of %arg6
>     %arg1: !llvm.ptr, // to memref<?x?xf64>
>     %arg2: !llvm.ptr, // to memref<?x?xf64>
>     %arg3: !llvm.ptr, // to memref<?xi64>
>     %arg4: !llvm.ptr, // to memref<?xi64>
>     %arg5: !llvm.ptr, // to memref<?xf64>
>     %arg6: !llvm.struct<(array<2 x i64>, array<3 x i64>)>) attributes {llvm.emit_c_interface} {
>   ...
> }
> ```

That last part is platform-specific and calling-convention specific, but I don't see a way to pass the calling convention into `mlir-opt` or `mlir-translate`. Specifically, it shows up here as a struct (as it should on x86_64), but on ARM64 one would expect this to be a pointer-to-struct.

**Comment by hameerabbasi** [comment]
cc @aartbik as the `sparse_tensor` dialect code owner.

**Comment by PeimingLiu** [comment]
I will take a look when I get time!

**Comment by hameerabbasi** [comment]
Just a light ping here, this is close to becoming blocking for us.

**Comment by aartbik** [comment]
MLIR does not guarantee a direct ABI between sparse tensors and the outside world. We guarantee such an ABI, however, through the assembler/disassembler operations, which provide a well-defined way to talk to outside systems (like PyTorch and JAX runtimes), both for sparse inputs and sparse outputs (return values).

Your command line:

```
"$LLVM_BIN/mlir-opt" "$SCRIPT_LOCATION/sddmm.mlir" \
    --sparsifier="enable-runtime-library=false" \
    --sparse-assembler="direct-out=false" \
    -llvm-request-c-wrappers \
    -o "$SCRIPT_LOCATION/sddmm_opt.mlir"
```

runs the sparse assembler too late. You will need to run this BEFORE the sparsifier in order to have effect (as written,  the pass has no impact at all). Swapping the two flags will resolve that issue. However, our assembler instructions for input expect a static size. When I change your code to, for example

```
// Define the function
func.func @sddmm_kernel(
    %arga: tensor<100x100xf64>,
    %argb: tensor<100x100xf64>,
    %args: tensor<100x100xf64, #CSR>
) -> tensor<100x100xf64, #CSR>
 ...
```

and then run the passes in the right order, a correct external method is generated that guarantees ABI clean semantics!

```
func.func @sddmm_kernel(%arg0: tensor<100x100xf64>, %arg1: tensor<100x100xf64>, %arg2: tensor<?xindex>, %arg3: tensor<?xindex>, %arg4: tensor<?xf64>, %arg5: tensor<?xindex>, %arg6: tensor<?xindex>, %arg7: tensor<?xf64>) -> (tensor<?xindex>, tensor<?xindex>, tensor<?xf64>) attributes {llvm.emit_c_interface} {
    %0 = sparse_tensor.assemble (%arg2, %arg3), %arg4 : (tensor<?xindex>, tensor<?xindex>), tensor<?xf64> to tensor<100x100xf64, #sparse>
    %1 = call @_internal_sddmm_kernel(%arg0, %arg1, %0) : (tensor<100x100xf64>, tensor<100x100xf64>, tensor<100x100xf64, #sparse>) -> tensor<100x100xf64, #sparse>
    %ret_levels:2, %ret_values, %lvl_lens:2, %val_len = sparse_tensor.disassemble %1 : tensor<100x100xf64, #sparse> out_lvls(%arg5, %arg6 : tensor<?xindex>, tensor<?xindex>) out_vals(%arg7 : tensor<?xf64>) -> (tensor<?xindex>, tensor<?xindex>), tensor<?xf64>, (index, index), index
    return %ret_levels#0, %ret_levels#1, %ret_values : tensor<?xindex>, tensor<?xindex>, tensor<?xf64>
  }
```
and then

```
llvm.func @_mlir_ciface_sddmm_kernel(%arg0: !llvm.ptr, %arg1: !llvm.ptr, %arg2: !llvm.ptr, %arg3: !llvm.ptr, %arg4: !llvm.ptr, %arg5: !llvm.ptr, %arg6: !llvm.ptr, %arg7: !llvm.ptr, %arg8: !llvm.ptr) attributes {llvm.emit_c_interface} {
...
```

**Comment by aartbik** [comment]
For our current use cases (PyTorch models) we use a closed FX graph export. JAX has the same constraints. As a result, all incoming tensors have static shapes. If you cannot cast your use case in that model, I think we will have to start supporting dynamic shapes with our assembler/disassembler as well.

**Comment by stellaraccident** [comment]
This may not be a "normal" thing.. We've had a problem before specifically on macos arm regarding register saving. Requires codegen to have the right target bits set.

I can't find the issue now, but it was related to x18 register reserves. https://stackoverflow.com/questions/71152539/consequence-of-violating-macoss-arm64-calling-convention

**Comment by aartbik** [comment]
Note that since there was some obvious confusion on this, I added a bit more documentation to the pass description.
Please let me know if that is more clear.

https://github.com/llvm/llvm-project/pull/96331

**Comment by hameerabbasi** [comment]
> Note that since there was some obvious confusion on this, I added a bit more documentation to the pass description.
> 
> Please let me know if that is more clear.
> 
> 
> 
> https://github.com/llvm/llvm-project/pull/96331

Thanks for the additional documentation! Yes, at some point I swapped the passes and it somehow started working with Dynamic shapes on amd64, and I believed you'd already added the functionality for dynamic shapes, but that it was still buggy. Turns out I was mistaken.

Static shapes indeed do not support my use case at the moment -- I'm happy to do what I can to help support dynamic shapes with a stable ABI, as it seems I'll need to get into the dialect anyway.

**Comment by aartbik** [comment]
> Static shapes indeed do not support my use case at the moment -- I'm happy to do what I can to help support dynamic shapes with a stable ABI, as it seems I'll need to get into the dialect anyway.

Can you please file a feature request for this, so we will track it properly? You can assign the request to me for now.

**Comment by hameerabbasi** [comment]
> Can you please file a feature request for this, so we will track it properly? You can assign the request to me for now.

I've filed #99442, but I don't have triage permissions.
